core/src/mast/merger/mod.rs

```diff
@@ -0,0 +1,429 @@
+                            })
+                            .collect::<Result<Vec<_>, _>>()?,
+                    ),
+                )
+                .expect("previously valid BasicBlockNode should still be valid")
+            },
+            MastNode::Dyn(_) => MastNode::new_dyn(),
+            MastNode::External(external_node) => MastNode::new_external(external_node.digest()),
+        };
+
+        // Decorators must be handled specially for basic block nodes.
+        // For other node types we can handle it centrally.
+        if !mapped_node.is_basic_block() {
+            mapped_node.set_before_enter(map_decorators(node.before_enter())?);
+            mapped_node.set_after_exit(map_decorators(node.after_exit())?);
+        }
+
+        Ok(mapped_node)
+    }
+
+    // HELPERS
+    // ================================================================================================
+
+    fn lookup_node_by_fingerprint(&self, eq_hash: &EqHash) -> Option<&(EqHash, MastNodeId)> {
+        self.node_id_by_hash.get(&eq_hash.mast_root).and_then(|node_ids| {
+            node_ids.iter().find(|(node_fingerprint, _)| node_fingerprint == eq_hash)
+        })
+    }
+
+    fn lookup_node_by_root(&self, mast_root: &RpoDigest) -> Option<&(EqHash, MastNodeId)> {
+        self.node_id_by_hash.get(mast_root).and_then(|node_ids| node_ids.first())
+    }
+
+    fn lookup_external_node_by_root(&self, fingerprint: &EqHash) -> Option<(EqHash, MastNodeId)> {
+        self.node_id_by_hash.get(&fingerprint.mast_root).and_then(|ids| {
+            let mut iterator = ids
+                .iter()
+                .filter(|(_, node_id)| self.mast_forest[*node_id].is_external())
+                .copied();
+            let external_node = iterator.next();
+            // The merging implementation should guarantee that no two external nodes with the same
+            // MAST root exist.
+            debug_assert!(iterator.next().is_none());
+            external_node
+        })
+    }
+}
+
+impl From<MastForestMerger> for MastForest {
+    fn from(merger: MastForestMerger) -> Self {
+        merger.mast_forest
+    }
+}
+
+// MAST FOREST ROOT MAP
+// ================================================================================================
+
+/// A mapping for the new location of the roots of a [`MastForest`] after a merge.
+///
+/// It maps the roots ([`MastNodeId`]s) of a forest to their new [`MastNodeId`] in the merged
+/// forest. See [`MastForest::merge`] for more details.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MastForestRootMap {
+    root_map: BTreeMap<MastNodeId, MastNodeId>,
+}
+
+impl MastForestRootMap {
+    fn from_node_id_map(id_map: MastForestNodeIdMap, roots: &[MastNodeId]) -> Self {
+        let mut root_map = BTreeMap::new();
+
+        for root in roots {
+            let new_id =
+                id_map.get(root).copied().expect("every node id should be mapped to its new id");
```

Comment:
My thinking here was that the constructor is private and presumably won't be exported in the future, so I thought we could take advantage of the guarantee that every id has an entry in the `MastForestNodeIdMap`.

I like to encode these kinds of guarantees as `expect`s instead of errors, because 1) they cannot happen (although refactorings are always a potential problem for these kinds of things) and 2) adding a dedicated error for something that the user never effectively sees to a public error type is adding "error variant" bloat to the public interface.

I'm not sure what the overall strategy is in our codebases regarding error handling. There tend to be a wide range of opinions regarding this, so if there is an established pattern, I'm happy to adapt :smile:.

Reply:
> My thinking here was that the constructor is private

Ah! My bad, I missed that.

> I'm not sure what the overall strategy is in our codebases regarding error handling. 

Nope, what you said makes perfect sense, and this is the way we want to do it in the VM for sure. 

(Note that there's tech debt in miden-vm and winterfell where we `expect()` on user input - the plan is to convert those to errors before mainnet)